### ISSUE TITLE
co_yield

### ISSUE THREAD

**Issue by unicomp21** [emscripten-core/emscripten#10991]
With emscripten using clang, should co_yield and generators "just work"?

**Comment by tlively** [comment]
Yes, they should. C++ generators are stackless, so all the interesting transformations should happen in the target-independent compiler frontend and we shoudn't need to change anything in the backend or on the Emscripten side. That being said, I don't think we've done any testing with generators. If you run into issues, please file an issue.

**Comment by unicomp21** [comment]
Awesome, thanks Thomas!

**Comment by unicomp21** [comment]
@tlively I'm a beginner on this topic, but I'm wondering.  If co_yield currently works within webassembly "as is", could it be used to implement co_await?  Wondering if it would be simpler than the binaryen implementation.

**Comment by unicomp21** [comment]
@kripken ^

**Comment by tlively** [comment]
co_await should also "just work" for the same reason as co_yield. Making C++ coroutines interoperate with JS async/await might be a little trickier, though. IIUC, we would need to implement some custom executor that coordinates with JS, but I'm a little fuzzy on the details of what that would look like.

> Wondering if it would be simpler than the binaryen implementation.

I assume you're thinking about whether Asyncify could be reimplemented with C++ coroutines somehow. It sounds attractive, but unfortunately it would not be a great idea. This is because Asyncify and coroutines are working at different layers of abstraction. In Asyncify, the language is not aware that execution is being paused so it needs to work on unmodified C and C++ code. C++ coroutines, on the other hand, are a language feature that programs have to opt into using via the new keywords in C++20. So in order to implement Asyncify in terms of coroutines, we would have to rewrite the code at some level to use the coroutine features. But even that wouldn't be quite enough because C++ coroutines only save the locals of the current function when it is suspended, but Asyncify needs to save the locals of every function on the stack. So in the end it's actually much simpler to just implement Asyncify as a transformation on the WebAssembly code in Binaryen.

**Comment by unicomp21** [comment]
Got it.  I can’t find any examples of co-await in webassembly, know if any?

**Comment by unicomp21** [comment]
https://github.com/emscripten-core/emscripten/issues/9316

@tlively @kripken

Starting to realize I may have been confused earlier, see above link.  Please correct me if I'm wrong.  The binaren coroutine support is more akin to boost fibers?  Whereas co_await is really a compiler supported concept?  If this is true, I should be able to create a promise implementation, using emscripten, which will support co_await?  Like has been done below?

https://devblogs.microsoft.com/cppblog/using-ibuv-with-c-resumable-functions/

https://github.com/jimspr/awaituv/blob/master/inc/awaituv.h#L300

**Comment by tlively** [comment]
Yes, that should work just fine, as long as it is not trying to interact with the JS event loop. That kind of integration would be more involved. For the particular example you shared, you would have to port libuv to WebAssembly, too. I'm not sure whether anyone has done that yet.

**Comment by unicomp21** [comment]
I don't need integration w/ JS event loop.

This is great!  I'll have to start play around with the compiler supported promise/future pattern in libuv, and see if I can get something working in webassembly.

**Comment by unicomp21** [comment]
@tlively @kripken what is @GorNishanov talking about here?  I don't understand the tail call part.

https://github.com/GorNishanov/await/issues/5

**Comment by GorNishanov** [comment]
To implement this feature: Add symmetric coroutine control transfer (open-std.org)<http://open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0913r0.html> efficiently compilers (clang, MSVC, probably GCC, I didn't check) rely on ability to do tail calls for synthesized coroutine parts.

Just double check that a code like this:

void coro$part2(struct X*);

void coro$part1(struct X* x)
{
    ...
    coro$part2(x);
}

when goes through emscripten and back will result in:

coro$part1(X*):
  ...
  jmp coro$part2(X*)

even at -O0

Cheers,
Gor
________________________________

**Comment by tlively** [comment]
Clang and Emscripten support WebAssembly tail calls with `-mtail-call`, but tail calls are not supported by any production WebAssembly engine that I know of.

**Comment by unicomp21** [comment]
@tlively i don't understand, meaning it will work?  or won't work?

**Comment by tlively** [comment]
I guess it should work but be inefficient.

**Comment by GorNishanov** [comment]
If a compiler, implements symmetric coroutine control transfer using tail calls, not honoring the tail call is a correctness issue. Since without tail calls, a thread stack could be easily exhausted.

It a platform does not support tail calls, a different codegen strategy has to be employed. Coroutine parts need to be no longer void, but, return the handle of the next coroutine to be executed and any resumption of the coroutine should be coded as:

// in some pseudocode form
void resume(coro$frame* coro)
{
  do
  {
      coro = call actual resume of coro
  }
  while (coro);
}

Clang/llvm relies on musttail to implement symmetric transfer. It will require non-trivial amount of work to implement symmetric transfer without relying on tail calls.

**Comment by unicomp21** [comment]
@kripken @tlively is this a binaryen thing?

**Comment by findmet** [comment]
Need help

**Comment by tlively** [comment]
> @kripken @tlively @sbc100 is this a binaryen thing?

No, what's at issue here is how Clang lowers coroutines and whether or not it tries to use mandatory tail calls. Since WebAssembly tail calls are not enabled by default, my guess is that Clang does not try to use them to lower coroutines. (If it did, that would be a bug). So coroutines should work fine but be slower than they would be if tail calls were enabled. It would be interesting to measure what the performance difference actually is for WebAssembly; it will probably be different than the difference would be on native architectures.

**Comment by unicomp21** [comment]
Has anyone tried using co_await on a chrome canary containing tail call support?

**Comment by unicomp21** [comment]
@GorNishanov could someone at MS try this out on an Edge browser canary build which supports wasm tail calls?
